src/normalize/selectioncompat.ts

```diff
@@ -13,105 +14,141 @@ export class SelectionCompatibilityNormalizer extends SpecMapper<
+      } else if (isBin(t) && isBinParams(t.bin)) {
+        return {
+          ...t,
+          bin: normalizeBinExtent(t.bin)
+        };
+      } else if (isLookup(t)) {
+        const {selection: param, ...from} = t.from as any;
+        return param
           ? {
-              condition: {
-                ...normalizeChannelDef(enc.condition),
-                test: normalizePredicate(enc.condition)
-              }
+              ...t,
+              from: {param, ...from}
             }
-          : {})
-      };
-    }
+          : t;
+      }
+      return t;
+    });
 
-    return {...rest, params, encoding};
+    return {...rest, transform};
   }
+
+  return spec;
 }
 
-function normalizeChannelDef(obj: any) {
-  const {bin, scale, selection, ...rest} = obj;
-  const {selection: param, ...domain} = scale?.domain || {};
-  return {
-    ...rest,
-    ...(bin ? {bin: isBinParams(bin) ? normalizeBinExtent(bin) : bin} : {}),
-    ...(scale
-      ? {
-          scale: {
-            ...scale,
-            domain: param ? {...domain, param} : domain
-          }
-        }
-      : {})
-  };
+function normalizeChannelDef(obj: any, normParams: NormalizerParams): ChannelDef {
+  const enc = duplicate(obj);
+
+  if (isFieldDef(enc) && isBinParams(enc.bin)) {
+    enc.bin = normalizeBinExtent(enc.bin);
+  }
+
+  if (isScaleFieldDef(enc) && (enc.scale?.domain as any)?.selection) {
+    const {selection: param, ...domain} = enc.scale.domain as any;
+    enc.scale.domain = {...domain, ...(param ? {param} : {})};
+  }
+
+  if (isConditionalDef(enc)) {
+    if (isArray(enc.condition)) {
+      enc.condition = enc.condition.map((c: any) => {
+        const {selection, param, test, ...cond} = c;
+        return param ? c : {...cond, test: normalizePredicate(c, normParams)};
+      });
+    } else {
+      const {selection, param, test, ...cond} = normalizeChannelDef(enc.condition, normParams) as any;
+      enc.condition = param
+        ? enc.condition
+        : {
+            ...cond,
+            test: normalizePredicate(enc.condition, normParams)
+          };
+    }
+  }
+
+  return enc;
 }
 
 function normalizeBinExtent(bin: BinParams): BinParams {
   const ext = bin.extent as any;
   if (ext?.selection) {
-    const {selection, ...rest} = ext;
-    return {...bin, extent: {...rest, param: selection}};
+    const {selection: param, ...rest} = ext;
+    return {...bin, extent: {...rest, param}};
   }
 
   return bin;
 }
 
-function normalizePredicate(op: any) {
+function normalizePredicate(op: any, normParams: NormalizerParams) {
   // Normalize old compositions of selection names (e.g., selection: {and: ["one", "two"]})
   const normalizeSelectionComposition = (o: LogicalComposition<string>) => {
-    return normalizeLogicalComposition(o, param => ({param} as any));
+    return normalizeLogicalComposition(o, param => {
+      const empty = normParams.emptySelections[param] || true;
```

Comment:
Should this use `??`? What happens if `normParams.emptySelections[param]` returns false?

Reply:
If it can never be false, we should use a `Set`.